pkg/services/user/userimpl/store.go

```diff
@@ -580,3 +584,30 @@ func setOptional[T any](v *T, add func(v T)) {
 		add(*v)
 	}
 }
+
+func handleSQLError(err error) error {
+	if isUniqueConstraintError(err) {
+		return user.ErrUserAlreadyExists
+	}
+	return err
+}
+
+func isUniqueConstraintError(err error) bool {
+	// check mysql error code
+	var me *mysql.MySQLError
+	if errors.As(err, &me) && me.Number == 1062 {
+		return true
+	}
+
+	// for postgres we check the error message
+	if strings.Contains(err.Error(), "duplicate key value") {
+		return true
+	}
+
+	var se sqlite3.Error
+	if errors.As(err, &se) && se.Code == sqlite3.ErrConstraint {
```

Comment:
Should we restrict this to a unique constraint error only? Looks like `ErrConstraint` would cover few other constraints (e.g. primary key, not null, etc).

```suggestion
	if errors.As(err, &se) && se.Code == sqlite3.ErrConstraintUnique {
```

Reply:
Yes so I was also looking into that one. `sqlite3.ErrConstraint` is an error code and `sqlite3.ErrConstraintUnique` is an error. Not sure how to check if it is of that type but I can look into it

Reply:
See [a9b2ac9](https://github.com/grafana/grafana/pull/101274/commits/a9b2ac9fd0538e2dd9b4c972bde79784e0020912)